File: ext/examples/chart.rs
use algebra::Algebra;
use ext::{
    chain_complex::{ChainComplex, FreeChainComplex},
    utils::query_module,
};
use sseq::charting::{SeqSeeBackend, SvgBackend, TikzBackend};

fn main() -> anyhow::Result<()> {
    ext::utils::init_logging()?;

    let resolution = query_module(None, false)?;

    let format = query::with_default("Output format (svg/tikz/seqsee)", "svg", |x| match x {
        "svg" | "tikz" | "seqsee" => Ok(x.to_string()),
        _ => Err(format!(
            "unknown format '{x}'; expected one of svg, tikz, seqsee"
        )),
    });

    let sseq = resolution.to_sseq();
    let products: Vec<_> = resolution
        .algebra()
        .default_filtration_one_products()
        .into_iter()
        .map(|(name, op_deg, op_idx)| (name, resolution.filtration_one_products(op_deg, op_idx)))
        .collect();

    let out = std::io::stdout();
    match format.as_str() {
        "svg" => {
            sseq.write_to_graph(SvgBackend::new(out), 2, false, products.iter(), |_| Ok(()))?
        }
        "tikz" => {
            sseq.write_to_graph(TikzBackend::new(out), 2, false, products.iter(), |_| Ok(()))?
        }
        "seqsee" => {
            sseq.write_to_graph(SeqSeeBackend::new(out), 2, false, products.iter(), |_| {
                Ok(())
            })?
        }
        _ => unreachable!(),
    }
    Ok(())
}

File: ext_py/examples/chart.py
#!/usr/bin/env python3
"""Write an SVG chart of the spectral sequence of a resolution to stdout.

Python port of ext/examples/chart.rs.
"""

import sys

import _query as query
from ext import sseq


def main():
    resolution = query.query_resolution()
    resolution.compute_through_stem(query.query_n_s())

    ss = resolution.to_sseq()
    products = [
        (name, resolution.filtration_one_products(op_deg, op_idx))
        for (name, op_deg, op_idx) in resolution.algebra().default_filtration_one_products()
    ]

    ss.write_to_graph(
        sseq.SvgBackend(sys.stdout),
        2,
        False,
        products,
        lambda _: None,
    )


if __name__ == "__main__":
    main()
